### PR TITLE
Let's improve support power timer display.

### PR DESCRIPTION
**Pull request by deniz1a**
Hello. I made a new PR. This is a simpler reimplementation of #7737. It organizes support power countdown timers in teams. Thank you.

### CODE REVIEW

**Comment by penev92** [comment]
Have you given up on the idea of getting the maximum name length in the constructor and using that? Because you have code commented out there.

**Comment by deniz1a** [comment]
I tried getting maxPowerNameWidth in constructor but couldn't do it. maxPlayerNameWidth could be in constructor but this way the texts align dynamically. So when a player with a long name doesn't build a support power, it won't be considered in alignment. But to do that this calculates maxPlayerNameWidth every tick.. So if dynamic alignment is not necessary, it could be done using the commented line in constructor.

Which one is better?

**Comment by Unit158** [comment]
Dynamic alignment is better in this case.

Can we get a better name for Line and change it to a struct please?

**Comment by deniz1a** [comment]
How should it be named? I thought line is the most fitting because that's what it represents. Any other suggestions?

Also, isn't class better than struct because it's passed by reference? Although it isn't passed to any method here, just added to a list. So it would probably make no difference? Should I change it to struct?

**Comment by deniz1a** [comment]
I could rename Line to Text...

**Comment by deniz1a** [comment]
And actually you can't change teams during game. Stance change in diplomacy window doesn't change player's team, only individual stances to other players... So I can move playersByTeam definition to the constructor.

**Comment by penev92** [comment]
This looks like a more complicated version of [this](https://github.com/penev92/OpenRA/blob/ba6b3b8d4315750056db5fa816078289b87c4d53/OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs).
What are the benefits of this one over that?

**Comment by deniz1a** [comment]
I don't know, I haven't tested that.

**Comment by deniz1a** [comment]
Can Draw() execute before the first Tick()?

**Comment by penev92** [comment]
OK, I'm going to be blunt.

You already had a PR about this that was way more complicated that it needed to be. Because of that a lot of developers spent *a lot* of time on it, including myself. *Then*, at one point, it reached a dead end. Instead of just closing it, I decided to give it a last chance, and spent around four hours of my time to rewrite it in a much simpler way. (Please keep in mind that even before I did so my opinion was that enough time was wasted on the PR by so many people for no gain.) Ultimately the PR was closed because you didn't want to use my suggested solution.

Now you are opening a new PR about the same thing (which is more than likely unfinished feature-wise). The code looks like a slightly more complicated version of the code I suggested for your previous PR, plus a (unnecessary?) `Tick()` method.
You say that you "haven't even tested" my suggested code. (?)

In light of that, I am going to -1 any solution that is more complicated than the one I proposed, unless you can explain how it is better than mine.

**Comment by deniz1a** [comment]
Why didn't you open a new pr with your version?

**Comment by penev92** [comment]
Because I don't consider this as something important, and definitely not something that justifies so many people wasting so much time on. And I still don't. So either we resolve this in a reasonable timeframe this time around, or I will close it.
Also because everyone else seems to be OK with the current state of the timers.

**Comment by deniz1a** [comment]
Are Tick() and Draw() methods synchronized? What happens when the game is paused? Tick() stops but Draw() continues, right? So it is better to put game state dependent code in Tick() and leave only drawing code in Draw() to minimize calculations.

**Comment by deniz1a** [comment]
If you don't consider this important and don't think this feature should be implemented, then why did you waste so much time, as you put it? Why did you write your own version and then not open a pull request?

**Comment by penev92** [comment]
> Instead of just closing it, I decided to give it a last chance

I figured that was the polite thing to do.
Don't get me wrong, I wanted to just close it and stop wasting everyone's time. And I definitely didn't want to spend 2 hours trying to figure out what your code does and 2 more simplifying it.

**Comment by deniz1a** [comment]
Ok, so you have the code that implements team grouping? Open a PR and if that's better than this one, it can be reviewed and merged...

**Comment by deniz1a** [comment]
I think this PR is finished and needs review. It does two things:
1) Separates SUPPORT_POWER_TIMER widget definition into ingame-player and ingame-observer. This makes it possible to define different bound offsets for observers who don't have support power icons.
2) Organizes support power countdown timers in player teams. It looks like this:
sptimer-simple

Power names and remaining times are aligned dynamically.
This map can be used to quickly test the timer display:
http://resource.openra.net/maps/3506/

Possible changes/improvements that could be done:
* Change ```class Text``` to ```struct``` as suggested by @Unit158.
* Calculate ```maxPlayerWidth``` once in the constructor for all players and not every tick for the players that have a support power.
* Calculate ```maxPowerWidth``` once in the constructor for all buildable support powers in current map or mod, instead of every tick. (I don't know how to do this.)
* Store ```playersByTeam``` in a data structure like Dictionary<> instead of querying it every tick. Or cache ```IOrderedEnumerable```. (I don't know how to do this.)

I can't think of anything else... OK, let's discuss!

**Comment by deniz1a** [comment]
Here are Visual Studio profiling results (.vsps files) for bleed and this PR:
https://drive.google.com/folderview?id=0B_dMvsHPqPZkfnJhR1Ryck44Szh4aDhJMjNHd2oySnpsbDNxdjd4UEc1MVhzY0NVcUE0RHM&usp=sharing

**Comment by abcdefg30** [comment]
Looks ok to me. :+1: / :white_check_mark:

**Comment by pchote** [comment]
Do we want this feature?  The reception has been luke-warm at best.

**Comment by deniz1a** [comment]
Doesn't this improve the timers considerably? Without team grouping and player names, I find the timers to be not that useful.

**Comment by pchote** [comment]
That is the open question which nobody has really asked yet.  It adds more details, but it also adds a lot more visual clutter to the screen.  If people don't feel that extra detail is needed, then the additional content could be taken as an overall negative instead of improvement.

**Comment by Unit158** [comment]
I'm going to day that player names are definitely a good thing as our colour validation is simply ass. Teams are a bit overkill but whatever.

**Comment by penev92** [comment]
Popular opinion has been that we do not want this.
Also :-1: until you can convince me this really needs `Tick()` *and* `Draw()` methods.

**Comment by GraionDilach** [comment]
I'm worried about that this can and will also run out of screen, increasing #7855 just even further.

**Comment by abcdefg30** [comment]
If we remove the `Team` stuff this is kinda just a bigger version of #6282.

**Comment by deniz1a** [comment]
I don't think team labels add that much to the length of the text. There are usually two teams... If that's a problem, team labels could be omitted if total number of teams exceeds a set number.

It is better to have separate Tick and Draw methods. It allows for separation of tasks. Tick does game state dependent logic, generates the actual text that will be drawn. Draw is only responsible for drawing on screen. It gets a list of texts and draws them. This has benefits for the following cases:
* When there is lag and Tick frequency is lower than Draw.
* When game is paused. Tick stops, Draw continues
* If in the future, Tick and Draw frequencies are decoupled, Draw could run at much higher rate than Tick. I think currently it runs with the same frequency as Tick (when there's no lag).

The other reason you listed is that this is more complicated than the version you implemented. If that is so, you should open a PR for it so we have a better implementation. It was nice of you to write that code to help me and I got some ideas from it. But what do you want now?

**Comment by deniz1a** [comment]
Apart from all that, the more important issue with support power timers is that drawing that text consumes 15% of CPU time according to the profiling results I posted above. If those numbers are correct, there is a problem with text drawing code. Maybe until that is fixed, countdown timers could be shown in a menu widget?

**Comment by deniz1a** [comment]
OpenRA.Graphics.SpriteFont.DrawTextWithContrast() consumes more than 14%.

**Comment by deniz1a** [comment]
I ran another profiling test and Draw() of SupportPowerTimerWidget uses 18.46% :worried:

**Comment by deniz1a** [comment]
I think Tick() and Draw() always run synchronized, so separating logic doesn't improve performance. But still it's better, right?

**Comment by deniz1a** [comment]
Widget's Tick() runs with LocalTick, which is separate from world tick, right? But it's not necessary to run the code in this widget's Tick() when world state hasn't changed. So how can I make it run with world tick instead of local tick?

**Comment by deniz1a** [comment]
Another issue with this code is that when game is paused, ready timers don't flash. Because colors are calculated in Tick() which is skipped if world tick hasn't changed. But I think this is not important.

**Comment by deniz1a** [comment]
Ok I think the game works like this:
There are three different frequencies or timesteps.
* **world.WorldTick** is the game progress. This stops when game is paused and slows down when there's lag. This is synchronized for all players. Runs every 40 ms or 25 times per second when there's no lag.
* **orderManager.LocalFrameNumber** is increased when the Tick() methods of all the classes are called. This is also 25 per second. I think this is not affected by game progress, so constant.
* **Game.RenderFrame** is the frame number of screen drawing. The frequency of this is as high as the computer can render or the frame limit value in settings menu, whichever is lower. This is increased when the Draw() methods of all the classes are called.

Then there's also network tick which has 1/3 the frequency of LocalFrameNumber. That's when packets are sent I think.
The frequency order is: 
RenderFrame > LocalFrameNumber > WorldTick > network tick

Is all this correct?

So, for this PR, we want to generate timer texts every WorldTick and draw them every RenderFrame which is when Draw() is called.

**Comment by deniz1a** [comment]
Although I just tested it and the RenderFrame frequency is around 250 fps and is not limited by frame limit setting...

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 1]
```diff
@@ -11,55 +11,146 @@
...
 		}
 
 		public override void Tick()
 		{
-			texts = powers.Select(p =>
+			if (lastTick == world.WorldTick)
+				return;
+
+			lastTick = world.WorldTick;
+			display = new List<Text>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Creating a new list every tick creates work for the garbage collector.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 2]
```diff
@@ -11,55 +11,146 @@
...
+				var teamDisplay = new List<Text>();
+				foreach (var player in team.Value)
+				{
+					var spiList = spManagers[player].Powers.Values
+						.Where(i => i.Instances.Any() && i.Info.DisplayTimer && !i.Disabled);
+
+					if (!spiList.Any())
+						continue;
+
+					var playerWidth = font.Measure(player.PlayerName).X;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: This won't change at runtime, so you can do it once in the constructor and then cache it.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 3]
```diff
@@ -11,55 +11,146 @@
...
+					if (playerWidth > maxPlayerWidth)
+						maxPlayerWidth = playerWidth;
+
+					var showPlayer = true;
+					foreach (var spi in spiList)
+					{
+						var powerWidth = font.Measure(spi.Info.Description).X;
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Likewise here - you know all the possible support power names before you start, so you can build a dictionary at startup instead of measuring everything every tick.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: If you precalculate the power and player names it becomes cheap to calculate the lengths, and you can then drop this memory churning caching completely.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 4]
```diff
@@ -11,55 +11,146 @@
...
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
+
+			playersByTeam = new SortedDictionary<int, List<Player>>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: What is the purpose of the `SortedDictionary` here?

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 5]
```diff
@@ -11,55 +11,146 @@
...
+			foreach (var team in spManagers.Keys.GroupBy(p => world.LobbyInfo.ClientWithIndex(p.ClientIndex).Team))
+			{
+				playersByTeam[team.Key] = new List<Player>();
+				playersByTeam[team.Key].AddRange(team);
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: You can do this in one line (and then drop the braces) with something like `playersByTeam[team.Key] = team.ToList()`.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 6]
```diff
@@ -11,55 +11,146 @@
...
+
+			font = Game.Renderer.Fonts[fontName];
+			yIncrement = font.Measure(" ").Y + 5;
+			padding = font.Measure("  ").X;
+
+			display = new List<Text>();
+			this.world = world;
 		}
 
 		public override void Tick()
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Widget ticking is decoupled from world ticking, so there is no benefit to implementing this in `Tick` compared with inside a `lastTick == world.WorldTick` check inside Render.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 4 above)
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: So the teams are returned in order. Normal dictionary returns items in undefined order, right?

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 7]
```diff
@@ -11,55 +11,146 @@
...
 		[ObjectCreator.UseCtor]
 		public SupportPowerTimerWidget(World world)
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			spManagers = new Dictionary<Player, SupportPowerManager>();
+			foreach (var tp in world.ActorsWithTrait<SupportPowerManager>().Where(tp => !tp.Actor.Owner.NonCombatant))
+				spManagers[tp.Actor.Owner] = tp.Trait;
```
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Should I cache a list of all support power instances that have DisplayTimer true? Instead of sp managers? Does that change during game?

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: That won't work, because the instances will change as the game progresses.  What you *can* do is query all the SupportPowerInfos and then store a dictionary of Info to width.

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: The list of SupportPowerInstance of a player changes during game? How do I get a list of all available powers then?

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Does this change during game?
https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/SupportPowers/SupportPowerManager.cs#L28

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: You can query them from the game rules.

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: > Does this change during game?

Yes.  Look at how it is used in that file, and you'll see why/how.

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Oh ok thanks.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 3 above)
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I can't get the names of all possible support powers... How can I do that?

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: `<Map rules>.<ActorInfos>.SelectMany(ai => ai.Traits.Where(t => t is SupportPowerInfo))`

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I used this and it threw TargetInvocationException:
```
maxPowerWidth = world.Map.Rules.Actors.Values
				.Select(ai => ai.Traits.GetOrDefault<SupportPowerInfo>())
				.Max(i => font.Measure(i.Description).X);
```

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: It will, yes. Your version is completely different to my hint above, and assumes that every actor type has exactly one support power. That assumption fails, so it throws.

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Ok, this doesn't crash but the list is empty:
```
maxPowerWidth = world.Map.Rules.Actors.Values
				.SelectMany(ai => ai.Traits.WithInterface<SupportPowerInfo>())
				.Max(i => font.Measure(i.Description).X);
```

**Comment by deniz1a** [comment]
Refactored the code. Minimized object creation. But I get an NRE when I sell allied tech center.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 8]
```diff
@@ -11,55 +11,144 @@
...
+				lastTick = world.WorldTick;
+				display.Clear();
+				alternateColor = Game.LocalTick % 50 > 25;
+				foreach (var team in playersByTeam)
+				{
+					needTeamLabel = true;
+					foreach (var player in team.Value)
+					{
+						var powers = spManagers[player].Powers.Values
+							.Where(i => i.Info.DisplayTimer && i.Instances.Any() && !i.Disabled).ToArray();
```
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Does this change while it is still being queried and cause the NRE?

**Comment by deniz1a** [comment]
Fixed NRE, added comments.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 8 above)
Comment (Unit158) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: no. the support power will be destroyed before it is removed from the list.

**Comment by deniz1a** [comment]
Looks finished to me, anything else that can be done?

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 9]
```diff
@@ -11,55 +11,148 @@
...
 
 namespace OpenRA.Mods.Common.Widgets
 {
 	public class SupportPowerTimerWidget : Widget
 	{
-		public readonly string Font = "Bold";
-		public readonly string Format = "{0}: {1}";
-		public readonly TimerOrder Order = TimerOrder.Descending;
+		[Translate] public readonly string NoTeamText = "No Team";
+		[Translate] public readonly string TeamText = "Team {0}";
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Exposing a direct C# format string to YAML is a bad idea. You sure there isn't a better way?

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Unfortunately not, but that's a discussion for our wider translation efforts and is way out of scope of this pr.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Alright then.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 10]
```diff
@@ -11,55 +11,148 @@
...
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick - lastTick > 9)
 			{
-				var font = Game.Renderer.Fonts[Font];
-				font.DrawTextWithContrast(t.First, new float2(Bounds.Location) + new float2(0, y), t.Second, Color.Black, 1);
-				y += (font.Measure(t.First).Y + 5) * (int)Order;
+				lastTick = world.WorldTick;
+				display.Clear();
+				alternateColor = Game.LocalTick % 50 > 25;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I see you carried this over but why not using 60 and 30 as magic values instead? Your changes turn ready texts from a balanced 50:50 blinking to one which prefers the player's color.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 11]
```diff
@@ -11,55 +11,148 @@
...
 		{
-			if (!IsVisible() || texts == null)
+			if (!IsVisible() || noPowers)
 				return;
 
-			var y = 0;
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick - lastTick > 9)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: This should be replaced with a (world.WorldTick % 10 == 0) check and lastTick should be removed then. The kind of optimization you're doing here should be done by the VM if needed.

Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I tried that and it seems to skip some ticks.. The countdown seconds look irregular, I don't know why?

Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Ah, I see, it must be a race condition when Tick is faster than rendering - in that scenario your code might be a better option indeed.

Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Another option worths a try is to use Game.LocalTick here to prevent that race condition.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 10 above)
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I think 50 and 25 were used because local tick frequency is 25 per second, if there is no lag. So this would ideally show player color and flashing color for one second each.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 12]
```diff
@@ -11,55 +11,150 @@
...
 		{
-			powers = world.ActorsWithTrait<SupportPowerManager>()
-				.Where(p => !p.Actor.IsDead && !p.Actor.Owner.NonCombatant)
-				.SelectMany(s => s.Trait.Powers.Values)
-				.Where(p => p.Instances.Any() && p.Info.DisplayTimer && !p.Disabled);
+			public Color PlayerColor;
+			public string Player;
+			public string Power;
+			public string RemainingTime;
+			public bool Ready;
```
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: Instead of having a separate Ready variable here, I could compare RemainingTime to zero time. I don't know which one is better.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs [hunk 13]
```diff
@@ -11,55 +11,150 @@
...
 		{
-			if (!IsVisible() || texts == null)
+			if (!IsVisible() || noPowers)
 				return;
 
-			var y = 0;
-			foreach (var t in texts)
+			// Generate texts once in every 10 WorldTicks,
+			// which is 0.4 seconds in game time.
+			if (world.WorldTick % 10 == 0)
```
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: The game cannot skip a world tick, right? Because that could cause this to not update often enough.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 10 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: "Ideally", but then you only check every "0.4 seconds" leading to 1.2 for one color and 0.8 for the other.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 13 above)
Comment (GraionDilach) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: >it must be a race condition when Tick is faster than rendering

Between your draws, the Ticks already pass. Either that, either the UI is separated from the world in this case.

Considering that the ready-text-blinking code uses Game.LocalTick instead, I'd assume that's much reliable than this. As I said it already.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 10 above)
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: I changed that, now text flashing is done in draw part of the method. So it will flash even when the game is paused.

File: OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs (hunk 13 above)
Comment (deniz1a) [inline OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs]: But that's different. Ready text blinking and countdown timer flashing are only visual indications. They don't depend on world state.

**Comment by pchote** [comment]
This has been modified since the original +1, so removing the needs +2 label (needs a full re-review).

**Comment by obrakmann** [comment]
:-1: I don't want this.

In fact, I'd rather see the support power timers reduced to your own (ie. just the icons), and if need be your allies. Why should my enemies know when my spy satellite launches?

**Comment by GraionDilach** [comment]
SpySat is a bad use-case. Knowing your enemy's Nuke/IonCannon is important. Not sure if we have a tag for hiding the countdown.

**Comment by deniz1a** [comment]
Actually I also think the countdown timers should be shown only for allies. But with this PR I was trying to improve the current display without changing any behaviour. Should I close this PR and open one that reduces timers to only those belonging to allies? Or this PR could still be useful for observers. So I could add that feature to this one.

**Comment by deniz1a** [comment]
Ok I removed timers that belong to players who are not allied with current player. Also removed player's own timers, they are already shown in icons.

**Comment by abcdefg30** [comment]
Sorry, but :-1: for only showing timers of allied players.

**Comment by deniz1a** [comment]
It's more logical and also improves gameplay I think.

**Comment by obrakmann** [comment]
OK, it seems we can't come to an agreement over this, so it's probably best to stay at the status quo.

**Comment by deniz1a** [comment]
I reverted the commit that hides enemy timers. I think there is an agreement that this would improve the game, the only part with disagreement was hiding enemy timers and I removed it.

**Comment by deniz1a** [comment]
I'm keeping my branch upto date with bleed, in case an agreement is reached that this is better than current timers. Code is here:
https://github.com/deniz1a/OpenRA/tree/countdown-simple

**Comment by GraionDilach** [comment]
Add something like a DisplayTimerToEnemies, set it false on GPS, true on AtomBomb and I think this is good then.

**Comment by deniz1a** [comment]
Should I open a PR with those changes? Why should atom bomb timer be shown to enemies but GPS not?

**Comment by GraionDilach** [comment]
Honestly I don't even understand why the GPS timer is even shown to the player, but meh. I'd consider the GPS as a support power (equal of Parabombs, Paratroopers or the Sonar Pulse) which does not majorly affect the game that much, compared to Chronosphere, Iron Curtain or the A-Bomb.

**Comment by deniz1a** [comment]
I think it affects the gameplay but those timers should be shown only to allies. I'll add DisplayTimerToEnemies and maybe open another PR.

**Comment by deniz1a** [comment]
I added ````DisplayTimerToEnemies```` to ````SupportPower````, so now it's configurable in yaml.

I think if there should be countdown timers in game, my countdown-simple branch is the way to do it. Even though the code is a lot more complex than the current implementation, there is no simpler way of doing it right. And despite the added logic, it is very efficient. It might even be faster to run than the current code, aside from any performance issues in font drawing.

In conclusion, I present to you this marvel of software engineering which improves the player experience significantly. The open question now is, if countdown timers are an important enough feature to justify maintaining this code. I think they are.

**Comment by pchote** [comment]
Please do not open another pr.  For the reasons that we have been over (and over and over...) across the last two PRs and the IRC channel this is not as straightforward as you claim. I think I can speak for everyone on the reviewing side by saying that we are not willing to reshash these arguments again.

In conclusion, no.  Lets not improve the support power display.  The timers are not important enough to justify the amount of hair that you are making us pull out in frustration over this.
